aten/src/ATen/native/AdaptiveAveragePooling3d.cpp

```diff
@@ -0,0 +1,315 @@
+#include <tuple>
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+
+namespace at {
+namespace native {
+
+namespace {
+
+inline int start_index(int a, int b, int c) {
+  return (int)std::floor((float)(a * c) / b);
+}
+
+inline int end_index(int a, int b, int c) {
+  return (int)std::ceil((float)((a + 1) * c) / b);
+}
+
+template <typename scalar_t>
+static void adaptive_avg_pool3d_out_frame(
+    scalar_t* input_p,
+    scalar_t* output_p,
+    int64_t sizeD,
+    int64_t isizeT,
+    int64_t isizeH,
+    int64_t isizeW,
+    int64_t osizeT,
+    int64_t osizeH,
+    int64_t osizeW,
+    int64_t istrideD,
+    int64_t istrideT,
+    int64_t istrideH,
+    int64_t istrideW) {
+  int64_t d;
+#pragma omp parallel for
+  for (d = 0; d < sizeD; d++) {
+    /* loop over output */
+    int64_t ot, oh, ow;
+    for (ot = 0; ot < osizeT; ot++) {
+      int istartT = start_index(ot, osizeT, isizeT);
+      int iendT = end_index(ot, osizeT, isizeT);
+      int kT = iendT - istartT;
+
+      for (oh = 0; oh < osizeH; oh++) {
+        int istartH = start_index(oh, osizeH, isizeH);
+        int iendH = end_index(oh, osizeH, isizeH);
+        int kH = iendH - istartH;
+
+        for (ow = 0; ow < osizeW; ow++) {
+          int istartW = start_index(ow, osizeW, isizeW);
+          int iendW = end_index(ow, osizeW, isizeW);
+          int kW = iendW - istartW;
+
+          /* local pointers */
+          scalar_t* ip = input_p + d * istrideD + istartT * istrideT +
+              istartH * istrideH + istartW * istrideW;
+          scalar_t* op = output_p + d * osizeT * osizeH * osizeW +
+              ot * osizeH * osizeW + oh * osizeW + ow;
+
+          /* compute local average: */
+          scalar_t sum = 0;
+          int it, ih, iw;
+          for (it = 0; it < kT; it++) {
+            for (ih = 0; ih < kH; ih++) {
+              for (iw = 0; iw < kW; iw++) {
+                scalar_t val =
+                    *(ip + it * istrideT + ih * istrideH + iw * istrideW);
+                sum += val;
+              }
+            }
+          }
+
+          /* set output to local average */
+          *op = sum / kT / kH / kW;
+        }
+      }
+    }
+  }
+}
+
+void adaptive_avg_pool3d_out_cpu_template(
+    at::Tensor& output,
```

Comment:
it is not a issue but just a comment, as it is also on the `at` namespace I think you can just use `Tensor` as you have use in other places in this file. but it is not something important .. just a comment

Reply:
Thank you. I like this kind of consistency and made changes.